lib/Conversion/TorchToTcp/Utils.cpp

```diff
@@ -130,6 +156,98 @@ Value broadcastShapeExceptDims(ConversionPatternRewriter &rewriter, Value input,
                                            axesAttr);
 }
 
+// the parameter values is expected to be an array of RankedTensorType tensors
+std::optional<SmallVector<Value>>
+broadcastManyToMatchShape(ConversionPatternRewriter &rewriter, Location loc,
+                          ValueRange values) {
+  if (values.size() <= 1) {
+    return values;
+  }
+  SmallVector<Value> ret;
+
+  int64_t maxRank = 0;
+  for (auto v : values) {
+    assert(isa<RankedTensorType>(v.getType()) && "assert 1");
+    auto t = cast<RankedTensorType>(v.getType());
+    if (t.getRank() > maxRank)
+      maxRank = t.getRank();
+  }
+
+  for (auto v : values) {
+    auto type = cast<RankedTensorType>(v.getType());
+    v = broadcastRankInLeadingDims(rewriter, v, maxRank - type.getRank());
+    ret.push_back(v);
+  }
+
+  // figure out what the shape should be for each dim
+  struct ShapeInfo {
+    Value value;
+    bool found = false;
+    int64_t static_value = 1;
+  };
+  SmallVector<ShapeInfo> shapes(maxRank);
```

Comment:
Is this `shapes` variable actually referring to a single shape?

Reply:
All of the input tensors need to be broadcast to the same shape, so this code is going through all of the tensors to try and determine what should be the resulting shape.  A dimension having a value of 1 means that it can be broadcast.  However, we can also find ourselves in the case where there are more than one tensor that have a non-1 value for the same dimension, in which case we should check that the two values are the same.

E.g. if you had tensors `1x5x2` and `3x1x2` then that would broadcast both of them to be `3x5x2`

Reply:
So, this `shapes` should refer to the final shape to which all the input tensors should be broadcasted to, right? In the example above, it must be `3x5x2` right?

My confusion is from whether this must be one shape or multiple shapes as the name `shapes` is referring to.